roseau/load_flow/models/voltage_sources.py

```diff
@@ -0,0 +1,91 @@
+import logging
+from collections.abc import Sequence
+from typing import Any, Optional
+
+from pint import Quantity
+
+from roseau.load_flow.exceptions import RoseauLoadFlowException, RoseauLoadFlowExceptionCode
+from roseau.load_flow.models.buses import Bus
+from roseau.load_flow.models.core import Element
+from roseau.load_flow.utils.json_mixin import JsonMixin
+from roseau.load_flow.utils.units import ureg
+
+logger = logging.getLogger(__name__)
+
+
+class VoltageSource(Element, JsonMixin):
+    """A voltage source.
+
+    The voltage equations are the following:
+
+    .. math::
+        \\left(V_{\\mathrm{a}}-V_{\\mathrm{n}}\\right) &= U_{\\mathrm{a}} \\\\
+        \\left(V_{\\mathrm{b}}-V_{\\mathrm{n}}\\right) &= U_{\\mathrm{b}} \\\\
+        \\left(V_{\\mathrm{c}}-V_{\\mathrm{n}}\\right) &= U_{\\mathrm{c}}
+
+    Where $U$ is the voltage and $V$ is the node potential.
+    """
+
+    def __init__(self, id: Any, n: int, bus: Bus, voltages: Optional[Sequence[complex]] = None, **kwargs) -> None:
+        """Voltage source constructor.
+
+        Args:
+            id:
+                The unique ID of the voltage source.
+
+            n:
+                Number of ports ie number of phases.
+
+            bus:
+                The bus of the voltage source.
+
+            voltages:
+                The voltages of the source. They will be fixed on the connected bus.
+        """
+        super().__init__(**kwargs)
+        self.connected_elements = [bus]
+        bus.connected_elements.append(self)
+
+        self.id = id
+        self.n = n
+        self.bus = bus
+        self.voltages = voltages
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}(id={self.id!r}, n={self.n}, bus={self.bus.id!r}, voltages={self.voltages!r})"
+
+    @property
+    def voltages(self) -> Sequence[complex]:
+        """The source voltages."""
+        return self._voltages
+
+    @voltages.setter
+    @ureg.wraps(None, (None, "V"), strict=False)
+    def voltages(self, voltages: Sequence[complex]) -> None:
+        if len(voltages) != self.n - 1:
+            msg = f"Incorrect number of voltages: {len(voltages)} instead of {self.n - 1}"
+            logger.error(msg)
+            raise RoseauLoadFlowException(msg, code=RoseauLoadFlowExceptionCode.BAD_VOLTAGES_SIZE)
+
+        if isinstance(voltages, Quantity):
+            voltages = voltages.m_as("V")
+        self._voltages = voltages
```

Comment:
I'm not sure those property are needed anymore: we used to have a `voltage` property because it was a result (for the `AbstractBus`) but now we can read the voltages of the bus attached to the voltage source.

Reply:
No the voltages can be updated between two load flows (for exemple if you want to simulate the tap uncertainty), so the `update_source_voltages` is still needed. It is overridden is `engine` to set internally the new voltages.